pkgs/applications/science/math/matlab/generic.nix

```diff
@@ -0,0 +1,105 @@
+{
+  version,
+  hash,
+  packages,
+}:
+
+{
+  lib,
+  stdenvNoCC,
+  autoPatchelfHook,
+  fetchurl,
+  buildFHSEnv,
+  makeDesktopItem,
+  mpm,
+
+  # Hash of downloaded files. You will need to change this if you include extra products.
+  withHash ? hash,
+
+  # Extra MATLAB products to install. E.g. Simulink or Optimization_Toolbox.
+  withProducts ? [ ],
+
+  withExtraPackages ? pkgs: [ ],
+}:
+
+let
+  products = [ "MATLAB" ] ++ withProducts;
+
+  # The MATLAB installation isn't 100% deterministic due to how the classpath.txt file is generated.
+  # But mpm download does appear to be deterministic. So create a fixed output derivation from the
+  # downloaded files, and then run the installer.
+  matlab-download = stdenvNoCC.mkDerivation {
```

Comment:
We could provide this as a fetcher (like `fetchFromMPM`) and use it to specify the `src` attribute of the `matlab` package.